File: linked-list/tasks/hasCycle.py
class Solution(object):
    """
    Input: head = [3,2,0,-4], pos = 1
    Output: true
    Explanation: There is a cycle in the linked list, where tail connects to the second node.

    通过两个指针一个快，一个慢，如果有环快指针肯定能追上慢指针
    """
    def hasCycle(self, head):
        slow = fast = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if slow == fast:
                return True
        return False

File: hasCycle/hasCycle.java

public class HasCycle {

    public static void main(String[] args) {
        System.out.println("reversrList start");
    }

    public static boolean hasCycle(ListNode head) {
        
      if (head == null) {
          return false;
      }
      // 通过两个快慢指针   
      ListNode fast = head.next;
      ListNode slow = head;
    
      while (fast != null && fast.next != null) {
          fast = fast.next.next;
          slow = slow.next;
        
          if (slow == fast) {
              return true;
          }
      }
      return false;
    }
}